src/main/java/org/breedinginsight/services/ProgramUploadService.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.breedinginsight.services;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.micronaut.context.annotation.Property;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.multipart.CompletedFileUpload;
+import lombok.extern.slf4j.Slf4j;
+import org.breedinginsight.api.auth.AuthenticatedUser;
+import org.breedinginsight.dao.db.enums.DataType;
+import org.breedinginsight.dao.db.enums.UploadType;
+import org.breedinginsight.daos.ProgramUploadDAO;
+import org.breedinginsight.model.Method;
+import org.breedinginsight.model.ProgramUpload;
+import org.breedinginsight.model.Scale;
+import org.breedinginsight.services.exceptions.DoesNotExistException;
+import org.breedinginsight.services.exceptions.UnprocessableEntityException;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Optional;
+import java.util.UUID;
+
+import org.breedinginsight.dao.db.tables.pojos.BatchUploadEntity;
+import org.breedinginsight.model.Trait;
+import org.breedinginsight.services.exceptions.UnsupportedTypeException;
+import org.breedinginsight.services.parsers.ParsingException;
+import org.breedinginsight.services.parsers.trait.TraitFileParser;
+import org.jooq.JSONB;
+
+import java.io.IOException;
+import java.util.*;
+
+@Slf4j
+@Singleton
+public class ProgramUploadService {
```

Comment:
Is the intention to put all of the file uploads into this service? Inventory, experiment, genotype, etc? Wondering if this file is going to become too big since everything is under a program?

Reply:
I could break it up similar to what we have to the ProgramController stuff. What do you think of having the one top level UploadController then a service for each upload type. I could also do a DAO for each type that could call to the generic ProgramUploadDAO with type abstracted away by the method names.

Reply:
Makes sense to break the controllers out (as well as make individual services), but not sure breaking out the DAOs is needed.  Wouldn't the service get the upload data into the right format for persisting, and be able to parse into the right domain model when reading?

Reply:
Renamed controller and service to reference trait upload and set pattern for future upload types. Left DAO as-is. Pushed changes.